[inductor] add subsystem to pattern matcher

avikchaudhuri opened the pull request:
Summary:
Running a toy example through `torch.compile(fullgraph=True, backend="inductor")` with default inductor config, I tried to see what passes are run in each of pre-grad, joint-graph, and post-grad phases by printing out the subsystem in `GraphTransformObserver`. However the subsystem showed up as None in a bunch of transforms that were run in each of those phases, so this PR adds some additional annotations.

Note that these annotations are probably not a complete set, since other transforms may run based on changes to the config that are not covered here.

Hopefully this doesn't change behavior. However, I did notice that bisecting relies on disabling various phases, which means that while before some passes would *not* be disabled (because their subsystem was `None`), now they would.

Test Plan: existing tests + manual test described in summary

Differential Revision: D83306676


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

jansel reviewed (changes requested):
Test failures?

eellison reviewed (commented):
nice ! but lets wait for test failures to be fixed..

avikchaudhuri committed:
[inductor] add subsystem to pattern matcher (#163922)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/163922

Running a toy example through `torch.compile(fullgraph=True, backend="inductor")` with default inductor config, we tried to see what passes are run in each of pre-grad, joint-graph, and post-grad phases by printing out the subsystem in `GraphTransformObserver`. However the subsystem showed up as None in a bunch of transforms that were run in each of those phases, so this PR adds some additional annotations.

Note that these annotations are probably not a complete set, since other transforms may run based on changes to the config that are not covered here.

Hopefully this doesn't change behavior. However, I did notice that bisecting relies on disabling various phases, which means that while before some passes would *not* be disabled (because their subsystem was `None`, now they would).

Test Plan: existing tests + manual test described in summary

Differential Revision: D83306676

avikchaudhuri commented on torch/_inductor/fx_passes/pre_grad.py:
```diff
@@ -1,5 +1,6 @@
 # mypy: allow-untyped-defs
 import copy
+import functools
```
Hmm, I'll do this in a separate PR.

qq: 
Looks like the bisector doesn't pick up something like:
```
with GraphTransformObserver(gm, "pre_grad_custom_pass"):
            config.pre_grad_custom_pass(gm.graph)
```

so instead I have to do:
```
GraphTransformObserver(gm, "pre_grad_custom_pass").apply_graph_pass(
            config.pre_grad_custom_pass
)
```

KP? @eellison

avikchaudhuri commented:
@pytorchbot merge
